### DISCUSSION TITLE
How to Modify Intellisense for ClassEnums? (auto-show "ClassName.")

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/roslyn#60058] [Q&A] [answered]
In order to create extendible Java-Like Enums, we are creating a class that attempts to emulate this.

The class has public statics to emulate the Enum members, and each returns an instance of Class Type.  So what it returns is an immutable registered instance of the Class itself...  emulating a smart-Enum  (has standard numeric Enum value, but also has attached meta data).

Example:
```c#
public class MapLayerName : ClassEnum    // ClassEnum is our base class that has members to make this work better.
{
    static readonly public MapLayerName Roads = new MapLayerName("Roads");
    static readonly public MapLayerName Rivers = new MapLayerName("Rivers");
}

//Usage Code:
MapLayerName layer = MapLayerName.Roads;
```

What we really want, is a strong-type in C# that accomplishes this, because classes are not-enum, and Intellisense knows this.   But such a type does not exist in C#, nor does it seem likely to ever exist in the foreseeable future.

Given that class-based Extendible Enums is our best option - we're trying to make this more elegant.

Here's the current obstacle, which makes Class-based Enums inferior to normal Enums.  When you type the following:

```C#
      ClassEnum enumInstance = ClassEnum.FirstValue;

      if (enumInstance ==        // <<=== Intellisense won't show me the desired "ClassEnum.{static members}"
```

We're looking for an easy way to make Intellisense recognize this as a special "ClassEnum" (even if through attribute) such that it'll show the "ClassEnum.{static members}" automatically in Intellisense.

Otherwise, the coder has to actually type "ClassEnums." directly, with no help from Intellisense.

And while we're at it, we'd like to filter out the static method "ReferenceEquals" in Intellisense for ClassEnums, as this pollutes the "enum values" that we want to show.

Ideally, whatever solution you propose is one that can be easily rolled out to Library users, without much issue.

This obstacle is main significant obstacle to using ClassEnums now, which is annoying to users of this construct.

**Comment by CyrusNajmabadi** [comment]
Moving to dotnet/roslyn as this is an intellisense question.

**Comment by CyrusNajmabadi** [comment, marked as answer]
To augument intellisense, you can MEF export a [CompletionProvider](https://github.com/dotnet/roslyn/blob/main/src/Features/Core/Portable/Completion/CompletionProvider.cs).  This will allow you to add whatever items you want here :)

**Comment by CyrusNajmabadi** [comment]
> And while we're at it, we'd like to filter out the static method "ReferenceEquals" in Intellisense for ClassEnums, as this pollutes the "enum values" that we want to show.

I don't believe we have any mechanism on this.  It's problematic as it's legal code, so if someone wants to write it, they may now find intellisense interfering with them.

**Reply by CyrusNajmabadi** [reply]
Actually, i think i'm wrong.  If you export a completion provider for this purpose and mark it as 'Exclusive', then i think we'll only show your items if you return any and will hide all others.

**Comment by CyrusNajmabadi** [comment]
Note: you should be able to add `/// <completionlist cref="MapLayerName"/>` to your `ClassEnum` type and have this automatically work.    Typing space after `==` should auto-select `MapLayerName`.

**Comment by najak3d** [comment]
I'll give this a try soon.   Do you have a recommendation for the best tutorial on creating a simple "CompletionProvider".  Basically, I just want to do this for any class the derives from the abstract base "ClassEnum" type.   If a class derives from "ClassEnum", then Intellisense will treat it more like an Enum type.

Starting uses case are when you type:  "var =", "var ==", and "var !=".   Those are the main use cases here.

And also whenever a method parameter name is of this type.

In all cases, the intellisense should show:   "{EnumClassName}" as the primary option, tap "TAB" and then "." and it shows you the static option names (anything static field/prop that is of type EnumClassName).

**Reply by CyrusNajmabadi** [reply]
> Do you have a recommendation for the best tutorial on creating a simple "CompletionProvider". 

Basically, just subclass that abstract class and override the methods :)

> Starting uses case are when you type: "var =", "var ==", and "var !=". Those are the main use cases here.

Generally we have that behavior (except the popping up on space):

![image](https://user-images.githubusercontent.com/4564579/157546074-ca78350d-7215-4dce-8fe3-be4cd9f9cab2.png)

You don't even need to hit tab.  You can either directly hit `.`, or you can see we automatically put your viable statics in the list directly.  So you can do:

```c#
if (layer == riv<tab>` 
```

and we'll complete out to `if (layer == MapLayerName.Rivers)`

**Reply by CyrusNajmabadi** [reply]
Note: if you add the `<completionlist` tag i mentioned above, then you get the space behavior you asked for:

![image](https://user-images.githubusercontent.com/4564579/157548591-fe8eacf9-d673-433d-bef6-d24148003f7d.png)

**Reply by najak3d** [reply]
In some contexts, I instead get all the way to typing "MapL" and it STILL doesn't know what I want.

The aim of intellisense control is "consistent" behavior that is equivalent to regular Enums.   I want there to be no doubt that the default option (after hitting Space) is always "MapLayerName".   

![image](https://user-images.githubusercontent.com/10914515/157549298-460c26d6-494c-4a80-bbd1-33846a6e6872.png)

I do admire how smart VS2022 Intellisense has become, the other day, was creating a Vector method called "Dot" and it entered the Dot equation for me!   Talk about AI.

But for "ClassEnums" I want Intellisense to treat identical how it treats Enums.   I don't want the users of ClassEnums to see ANY difference at all, if possible -- other than the benefits (the added meta data/smarts that can now be associated with each enum value, as well as the ability to Add new Enum values, easily).

**Reply by CyrusNajmabadi** [reply]
> In some contexts, I instead get all the way to typing "MapL" and it STILL doesn't know what I want.

Don't type prefixes.  Type things like `mln` to get hump-matching.  It's much faster and works better for long names with common prefixes.

**Reply by CyrusNajmabadi** [reply]
> But for "ClassEnums" I want Intellisense to treat identical how it treats Enums. I don't want the users of ClassEnums to see ANY difference at all, if possible

Again, that is our behavior if you use the `<completionlist` tag.  It's literally the same completion provider on our end.

**Reply by CyrusNajmabadi** [reply]
You can see that here: https://github.com/dotnet/roslyn/blob/main/src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs

It is literally called 'Enum *and* CompletionTag' completion provider.  Meaning we treat enums and `<completion` tag cases practically the same way in terms of experience.

**Reply by najak3d** [reply]
Maybe I'm misunderstanding something.  I just made my code look like this, but there is not noticeable difference for me in how the Intellisense works.

```c#
	/// <completionlist cref="MapLayerName"/>
	public partial struct MapLayerName : IComparable, IComparable<MapLayerName>, Enums.IStructEnum
       {... }
```

**Reply by CyrusNajmabadi** [reply]
You put that tag on ClassEnum

**Reply by najak3d** [reply]
> You put that tag on ClassEnum

But ClassEnum is a base class, maybe in another's library.  ClassEnum won't know about all of the variations that might exist, so don't think it's a viable solution to have the tag on ClassEnum.   I'm confused as to why this tag should work on the base Class, but not directly upon each derivation of ClassEnum?

I'm aiming to make this into a very-easy-to-use alternative to Enums -- not quite as efficient as standard Enums, but packs more functionality, and most importantly -- is extendible.   I'd like these to be a first-class easy-to-use construct, and thus, they need to look/feel almost exactly like normal Enums (with perks).

**Reply by CyrusNajmabadi** [reply]
> ClassEnum won't know about all of the variations that might exist, so don't think it's a viable solution to have the tag on ClassEnum. I'm confused as to why this tag should work on the base Class, but not directly upon each derivation of ClassEnum?

How would intellisense know any of those types when you do: 

```c#
     ClassEnum enumInstance = ClassEnum.FirstValue;

      if (enumInstance ==<space>
```

?

**Reply by CyrusNajmabadi** [reply]
> But ClassEnum is a base class, maybe in another's library. ClassEnum won't know about all of the variations that might exist, so don't think it's a viable solution to have the tag on ClassEnum.

I don't see why not.  Intellisense can only select one item, so it's not like you have any options when there are multiple subclasses.  You're asking for enum-like behavior, but enums can work this way precisely because there is only one enum type.

> I'd like these to be a first-class easy-to-use construct, and thus, they need to look/feel almost exactly like normal Enums (with perks).

I've laid out the options with intellisense.  You can use that comment construct provided, or you can create a completion provider for your library with the behavior you want :)

**Comment by najak3d** [comment]
I see I've miscommunicated "ClassEnum" -- this is different than my old "base class" which was a struct, and had Members on it already.

Now that you've reasoned me into the "ClassEnum" concept vs. the "StructEnum" -- now I can actually create a generic non-specific base class which will have common logic to be shared by ANY/ALL ClassEnums.   I would actually call this class "ClassEnum" and it would be abstract, and re-useable in any domain.  Fully generic.

MapLayerName then would derive from ClassEnum, and adds domain specific Enum values -- i.e. the MapLayers.

Then, other assemblies which use this assembly for MapLayerName, will want flexibility to add new MapLayerNames... and so they derive from the base MapLayerName and add members (and maybe more methods if they want, which only apply to their derivation).

The idea is that the fully generic "ClassEnum" abstract class will exist in a shared Utility Library -- and will NEVER know anything about the types that derive from it.

===
Ideally, ClassEnum itself would possess some sort of Attribute, or we have some sort of "CompletionProvider" which follows the strict rule that "any class which derives from ClassEnum will be treated like an Enum in Intellisense".   And thus we alleviate work for anyone wanting to use the ClassEnums.

I might instead call "ClassEnum" as "SmartEnum" for marketing reasons, or maybe "OpenEnum" to follow other recent trends of prefixing with "Open" -- which also describes them to a point  (open to extensions and added functions).

**Reply by najak3d** [reply]
Whatever we're building here -- IMO, should probably be a part of .NET framework.  
 
Many standard (non-intelligent, non-extentible) Enums can be replaced with these, as well as the rampant use of raw String-Keys.

**Examples:**
In our map app, we have layers -- these will be SmartEnums.   Our map has Modes -- again, SmartEnums. 

A game we're working on has ResouceTypes -- will be Smart Enums.   It has Labor types, building Types, Professions, etc, etc ... all will become "Smart Enums" -- and Smart Enums are not limited to "only the stuff hardcoded" -- they still work just fine if you did need to pass in raw strings - and can intermingle it with the rest of the Enum-values that are hardcoded.   They are very flexible -- yet when it boils down to it -- still gives you the elegant safety of using a raw Enum, for when it comes to "typing the name correctly" - the compiler ensures you are accurate, and if you refactor a name, the IDE will change it in all places.   

I'm very glad that you helped me see the light that ClassEnums a better option than StructEnums.   It's really opened the door for making these quite good.

**Reply by CyrusNajmabadi** [reply]
> Then, other assemblies which use this assembly for MapLayerName, will want flexibility to add new MapLayerNames... and so they derive from the base MapLayerName and add members (and maybe more methods if they want, which only apply to their derivation).

i'm not sure how you expect intellisense to work here.  Sounds like you have a base type with potentially many derived types.  What would intellisense preselect after `==` in this case?  why would any of the other types be any more suitable than the base type?

**Reply by CyrusNajmabadi** [reply]
> Ideally, ClassEnum itself would possess some sort of Attribute, or we have some sort of "CompletionProvider" which follows the strict rule that "any class which derives from ClassEnum will be treated like an Enum in Intellisense". 

Sure, writing your own provider for this seems reasonable.

**Reply by CyrusNajmabadi** [reply]
>  IMO, should probably be a part of .NET framework.

I'm really not sure what you are proposing be part of the .net framework.  It sounds like you have a normal inheritance hierarchy here with base types, derived types, and static instance of some of the types in your inheritance hierarchy.  That seems like just normal programming already enabled by teh standard .net framework and C# patterns.  What else are you looking for in the framework to assist here?

**Reply by najak3d** [reply]
I'm just stating that the OpenEnums concept is a very useful concept.    As with many other generically useful classes, they made their way into the "System" namespace - and that's all I'm talking about.   I'd like to see "System.OpenEnum" or something like it become a standard available to everyone - vs. just those who find/use a specific library.  

Imagine if TimeDate were removed from System namespace -- that would be a detriment.  Or if System.Linq namespace were removed, and turned into a 3rd party library, not shipped with .NET.   In the same way, I have a belief that the OpenEnum concept is of a similar (generically very useful) nature.   I think .NET should ship with such a concept built-in.   I could be wrong of course.

**Reply by najak3d** [reply]
The CompletionProvider should simply look at the actual class of the variable at hand, and start by showing that class name.  
From then on, it can behave normally, with exception of hiding the "IsReferenceEquals()" static method.

I suppose from what you are implying is that there is nothing pre-made that already does this, and so leaves it to me.   I've never written one of these before; how involved do you think it'll get?  (how many SLOCs approx should I expect to write?)   Just looking for your ballpark estimate.

**Reply by CyrusNajmabadi** [reply]
> I'm just stating that the OpenEnums concept is a very useful concept. As with many other generically useful classes, they made their way into the "System" namespace - and that's all I'm talking about. I'd like to see "System.OpenEnum" or something like it become a standard available to everyone - vs. just those who find/use a specific library.

Create an API proposal at dotnet/runtime for this.  like with csharplang design there are specific processes to follow to make these things happen.  They won't happen on their own.  thanks!

**Reply by CyrusNajmabadi** [reply]
> Or if System.Linq namespace were removed, and turned into a 3rd party library, not shipped with .NET. 

That would be completely fine with me.  Were we to do things over again, i imagine most things would come in as separate dlls (potentially with packs for easy distribution).

**Reply by CyrusNajmabadi** [reply]
> The CompletionProvider should simply look at the actual class of the variable at hand, and start by showing that class name.
From then on, it can behave normally, with exception of hiding the "IsReferenceEquals()" static method.

This is how it already works.  Just add the `<completionlist` tag to your derived types.  I've mentioned this a few times.

> I suppose from what you are implying is that there is nothing pre-made that already does this,

I mean... i think this shoudl already work, and trying it out locally it does (i even included pictures and whatnot).  This is what i get, for example, after hitting space:

![image](https://user-images.githubusercontent.com/4564579/157714755-c21af375-e032-4fee-87c9-84cc9f894dd0.png)

**Reply by najak3d** [reply]
How about System.Collections.Generic?   Wouldn't it be a pain if there were 100 different versions of equally-used List<T> classes, that were incompatible?      There is good reason for packing together "the standard stuff" -- so that it becomes "standard".   You can always expect that List<T> is available anywhere/everywhere in .NET -- and so is often used.   Or the ILIst<T> interface -- interfaces are only helpful if they are standard.   If 19 other libs have their own IList<T> - what good is that?

So I find it odd that you'd suggest the entire framework be divided into a-la-carte pieces, packaged more like 3rd party libs -- not directly from Microsoft -- thereby giving rise to "many versions" of each.

**Reply by CyrusNajmabadi** [reply]
> and so leaves it to me. I've never written one of these before; how involved do you think it'll get? (how many SLOCs approx should I expect to write?) Just looking for your ballpark estimate.

Our own code for this is around 300 lines.  So roughly that?  It really depends on what you're doing and how well the concepts you need map to the concepts the roslyn API exposes.  If it's close, then less code.  If it's far, then you'll need more.

**Reply by najak3d** [reply]
Your example with SomeEnumType is too simple..  Introduce other classes that start with "S" and "So"...  and then put "SomeEnumType" into another namespace brought in via using, but make sure the local namespace has other types that start with "Some" - then see how Intellisense behaves.

Your example works because there is literally little else for Intellisense to guess here.

What we need is a CompletionProvider that simply says "guess the actual ClassName FIRST, always, no matter the context".   That's all we need for it to do.

**Reply by CyrusNajmabadi** [reply]
> How about System.Collections.Generic? 

I would be fine with the collections coming from a disparate dll/package.

**Reply by najak3d** [reply]
How does one insert a CompletionProvider into a Solution?   Then how is this done for customers of your library (automatically)?

**Reply by CyrusNajmabadi** [reply]
Still works exactly as i mentioned:

![image](https://user-images.githubusercontent.com/4564579/157716096-f403db0c-acf4-4b1d-833f-7d5b9d3fd2f1.png)

![image](https://user-images.githubusercontent.com/4564579/157716159-75756bf2-b458-4d49-9aeb-06fef0a661ec.png)

After typing `<space>` or `S` we still keep all this preselected *just like an enum*.  Why? because it's literally the same code.  I linked you to that. :)

If you type `So` we will switch to teh exact match:

![image](https://user-images.githubusercontent.com/4564579/157716284-00d08860-be5a-46a7-b72a-cb53880364c1.png)

But we do that with enums as well.

**Reply by najak3d** [reply]
> > How about System.Collections.Generic?
> 
> I would be fine with the collections coming from a disparate dll/package.

I'm surprised at this view.   It would be like saying "I don't care for there to be standards" -- think of hardware, where standards allow everyone to use the same parts, because all parts conform to the standards.

If you have no standard "IList<T>" -- this would cause a sort of compatibility Hell.  For example, if your lib uses "IList<T>" while an assembly you are using is returning to you "IListMutant<T>" - you can't deal with it, without having many versions of List<T> within your own app.    Madness.    I'm surprised you are OK with that sort of ecosystem.

**Reply by CyrusNajmabadi** [reply]
> I'm surprised at this view. It would be like saying "I don't care for there to be standards" 

I never said that.  You are mistaking distribution mechanism for standardization. 

Regardless, if you think your idea has merit, open a proposal at dotnet/runtime.  This repo is not involved in the process of adding types to the .net standard libraries.

> I'm surprised you are OK with that sort of ecosystem.

I didn't describe such an ecosystem.  That's a strawman :)

You can have standards that are defined, but then have different impls of that standard.  Indeed, that's very often a *common* way that things are done.  Consider how in C/C++/Java tehre may be standards around their APIs, but there can be literally dozens of different implementations taht you can get from varying sources.

**Comment by najak3d** [comment]
I'm testing this from within my own project, and even created "MapLayerName2" of type class, instead of struct, to make sure this wasn't a struct thing..   and intellisense is so confused that it shows NO OPTIONS after hitting space for this instance.

![image](https://user-images.githubusercontent.com/10914515/157717122-e556d5f8-7217-4f78-a63b-0b83f6f39a1a.png)

While if I make a like-named "MapLayerNameEnum" of type "enum", it works perfectly, everywhere:

![image](https://user-images.githubusercontent.com/10914515/157717462-a1c76d09-99fd-41d3-9fc0-ce542893f3d9.png)

**Reply by CyrusNajmabadi** [reply]
![image](https://user-images.githubusercontent.com/4564579/157718342-3df20de3-2fb7-48a4-886a-4d36669b603d.png)

**Reply by CyrusNajmabadi** [reply]
![image](https://user-images.githubusercontent.com/4564579/157718472-9fff3d37-31fc-453f-90c6-252371103624.png)

**Reply by CyrusNajmabadi** [reply]
this is after hitting space in both cases.

**Reply by CyrusNajmabadi** [reply]
Again, there is very little difference between the behavior here.  The logic for enums is here:

https://github.com/dotnet/roslyn/blob/bff4ff14f69b41d09044de60bf1cab7ed4a7e1c7/src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs#L181-L200

The logic for classes that have the doc comment tag is here, just a few lines below: 

https://github.com/dotnet/roslyn/blob/bff4ff14f69b41d09044de60bf1cab7ed4a7e1c7/src/Features/CSharp/Portable/Completion/CompletionProviders/EnumAndCompletionListTagCompletionProvider.cs#L205-L240

**Reply by najak3d** [reply]
Here's what I see from inside my project:

![image](https://user-images.githubusercontent.com/10914515/157719375-ededc2dd-f0cd-408c-8d34-fe772af7eff5.png)


Initially it shows NOTHING...  I type "S" and it shows "SomeClass", not "SomeEnum"..... and keeps with it.

**Reply by najak3d** [reply]
I've been using StructEnums (similar concept) for years, and never has it reliably (or ever?) shown me the StructName as an option in Intellisense-- which is why I raised the issue here.

**Reply by CyrusNajmabadi** [reply]
You are not following the instructions i gave: https://github.com/dotnet/roslyn/discussions/60058#discussioncomment-2333190

**Reply by CyrusNajmabadi** [reply]
> I've been using StructEnums (similar concept) for years, and never has it reliably (or ever?) shown me the StructName as an option in Intellisense-- which is why I raised the issue here.

I understand.  And i've explained several times in the thread how to get that behavior by default.  If you don't want to follow those steps though, you can also just write your own completion provider (though it seems like it would be reinventing the wheel here). :)

**Reply by CyrusNajmabadi** [reply]
i just tried with your code and my instructions and i get the behavior you are asking for:

![image](https://user-images.githubusercontent.com/4564579/157720221-d56fed05-b66e-45ff-a7ee-8506c8c94ced.png)

**Reply by najak3d** [reply]
The busier the namespace/complexity -- the more likely it is to show nothing.    And adding/subtracting the "completionlist" tag often changes nothing...   it behaves like there is also some AI playing a role here, which causes behavior to vary.

I'm still getting NOTHING after hitting space, OFTEN -- but at times it does show "SomeEnum" as I want it to...  It's unreliable, unlike Enum types which are reliable.

**How does one insert a modified CompletionProvider for an entire solution?  And also for a client's solution?**

![image](https://user-images.githubusercontent.com/10914515/157723867-a3a8da5a-b4e2-4ecd-aa8a-6e34373eaf4d.png)

**Reply by najak3d** [reply]
To reproduce the failure, it's vital to make sure that there are at least several other viable options in the namespace -- more complexity/options is what seems to mess up the intellisense results.   While for Enums, no amount of complexity messes it up.

I'd like to keep the default Provider, but just add a single branch into the "class handling" - where it checks for if there is a derivation from "System.OpenEnum" type, and if so, then is strict on it's results for when user hasn't yet entered anything -- just tapped "SPACE").   That's the ONLY change I need.

**Reply by najak3d** [reply]
Just to demonstrate the unreliability I'm seeing here ... I played around some more, and for Enum2 the Intellisense came back:

![image](https://user-images.githubusercontent.com/10914515/157724774-69c1f99f-cb3a-4905-9338-82ab354a2595.png)

However, if I dink around with this code, then come back to this version -- later, it'll fail.

**Reply by CyrusNajmabadi** [reply]
![image](https://user-images.githubusercontent.com/4564579/157724453-46cbe4a1-ea6b-4097-af04-1f7aa8a03186.png)

![image](https://user-images.githubusercontent.com/4564579/157724527-637c161f-4f70-4c64-92af-13ac262c3079.png)

> The busier the namespace/complexity -- the more likely it is to show nothing

There should be nothing related to the namespace that affects this.  You can see the code that controls this in the code i linked.

> And adding/subtracting the "completionlist" tag often changes nothing... 

It changes the behavior to behave like an enum or not.  Again, you can see the code that in the code i linked :)

If you remove it, you won't get this behavior.

> it behaves like there is also some AI playing a role here, which causes behavior to vary.

There is no AI at play here.  

> How does one insert a modified CompletionProvider for an entire solution? And also for a client's solution?

I believe it's packaged up like any other roslyn plugin.  We likely have docs for this somewhere around here.  But it doesn't seem like it would be necessary here as the tag approach suffices. :)

**Reply by CyrusNajmabadi** [reply]
> Just to demonstrate the unreliability I'm seeing here .

*Again* (for like the 10th time) :)  this is the same codepath as the enum completion codepath.  If you're seeing unreliability based on structure of the project, then you'd see that as well with the enum path.  It's possible there's some bug here, but in that case, we can just address it if you have a code repro that shows when it's not showing up.

**Reply by CyrusNajmabadi** [reply]
> It's unreliable, unlike Enum types which are reliable.

it's literally teh same code.  if you have a case where it's not reliable, please provide the code that demonstrates this.  We can trace through to see what might be happening?  Note: the only thing that i can think that could possibly be is that completion is given a time budget.  So, if for some reason your machine is bogged down and taking too long, then we'll cancel the completion list.  But the budget is super high.  So if you're hitting that, something is seriously bugged.

**Reply by CyrusNajmabadi** [reply]
> To reproduce the failure, it's vital to make sure that there are at least several other viable options in the namespace

I've tried this in roslyn itself (which has tens of thousands of types).  The behavior works as expected. If you can make a repro project where it doesn't work, please do so and link us to it.

**Reply by najak3d** [reply]
I finally got it to work in our project using a "Class Enum" vs. StructEnum.   I've never gotten this to work for StructEnum...  But for ClassEnum it finally worked -- but as I said, was unreliable.   BUT -- maybe if I stop dinking around with the tag, then it'll stabilizer, and work consistently for the ClassEnum. 

I'll stop adding/subtracting the tag, and just let it sit there unchanged -- and see if it remain reliable.  I imagine it will, based upon the code review.

**Reply by CyrusNajmabadi** [reply]
> I'd like to keep the default Provider, but just add a single branch into the "class handling" - where it checks for if there is a derivation from "System.OpenEnum" type, and if so, then is strict on it's results for when user hasn't yet entered anything -- just tapped "SPACE"). That's the ONLY change I need.

That's the provider i linked to, and that's the behavior it supports.  I'm not sure what else to tell you.  Trying this out myself (even in a large solution) works fine.  If you can provide a case where it doesn't, i can definitely take a look to see what's up.  But absent that, i'm not sure what to tell you.

**Reply by najak3d** [reply]
Is it expected behavior for this tag to fail for structs?

	/// <completionlist cref="MapLayerName"/>
	public partial struct MapLayerName : IComparable, IComparable<MapLayerName>, Enums.IStructEnum  { ...}

**Reply by CyrusNajmabadi** [reply]
Works fine for structs:

![image](https://user-images.githubusercontent.com/4564579/157726043-193156d1-abb9-41ef-8391-c8ca3fc7bd64.png)

![image](https://user-images.githubusercontent.com/4564579/157726075-98106468-9bb3-47b2-8652-a06bc9476e32.png)

**Reply by CyrusNajmabadi** [reply]
![image](https://user-images.githubusercontent.com/4564579/157726213-38ca6af8-c0d9-4f4c-9bf4-cd82bae02e65.png)

**Reply by najak3d** [reply]
I found the culprit for MapLayerName.  It has a Newtonsoft.Json.JsonConverter attribute that causes it to fail.
```c#
	[Newtonsoft.Json.JsonConverter(typeof(Enums.StructEnumHelper<MapLayerName>.JsonConverter))]
	/// <completionlist cref="MapLayerName2"/>
	public partial class MapLayerName2
	{
		static public MapLayerName2 One { get; private set; } = new MapLayerName2();
	}
```

If I remove this attribute, then it works just fine.   Intellisense then works like an Enum.

Not sure why this attribute would mess it all up -- but it's 100% consistent in my tests.

**Reply by CyrusNajmabadi** [reply]
`warning CS1587: XML comment is not placed on a valid language element`

Your doc comment is not legal there.  They must go before the entire construct (including attributes).

**Reply by najak3d** [reply]
Well that was it --  "completionlist" tag must come before the attributes.   Since the attribute works regardless of order, I was unaware that a "///" tag was order dependent.   Well there we have it.

"completionlist" should do the trick.   It does behave "unreliably" for a large 300K SLOC app, but this "settles out", especially if you Compile/run it... then it clear it up, and behavior then stays stable.

Because it was failing for MapLayerName (due to this attribute) - I just figured it was deficient.   Apparently it's not.

Thank you for sticking with me on this.   This is very good news.  .NET wins again.

**Reply by CyrusNajmabadi** [reply]
Is the unreliability during, or near after, solution load?  If so, this is generally by design as we are operating in a limited mode because we don't actually understand all symbols yet.  This tends to be pretty short lived, but is dependent on machine configuration.

**Reply by najak3d** [reply]
It's now 100% what we needed.  Even works like this for method parameters.

![image](https://user-images.githubusercontent.com/10914515/157732523-7a69b909-be97-43b8-9cdf-165c764a481e.png)


I will be able to upgrade StructEnums to ClassEnums without a hitch.   The result should be "near Enum equivalent" behavior.

Last nit, is hiding "ReferenceEquals()" - I assume this would require an actual replacement of the existing CompletionProvider.

![image](https://user-images.githubusercontent.com/10914515/157732803-35626b0a-d4f9-4643-8378-c07b93535be8.png)

**Reply by najak3d** [reply]
Most people get away from "raw string key" by defining them statically -- so that you have intellisense to ensure you enter the static property name correctly (or Const).    The pit-fall here, is that still "it's just a string".

With OpenEnums - many of these texts can pack more meta data with them, and not just be a string value.   For example, text messages might also come with a Style modifier and color code, and "priority level", etc..    So now you can have a whole list of these strings called out like an "enum value" but packed with meta data  -- it's no longer "just a string", and yet is as easy to work with as an Enum Value.

The improvement of this concept, which you helped with, excited me a lot.   And I will say it again -- I'm surprised that something like this hasn't become a standard part of .NET already (as it is with Java Enums).

**Reply by najak3d** [reply]
> Is the unreliability during, or near after, solution load? If so, this is generally by design as we are operating in a limited mode because we don't actually understand all symbols yet. This tends to be pretty short lived, but is dependent on machine configuration.

It is something like that.  It happens whenever I just modify stuff, or add/subtract the XML tag.   It can last for over a minute, at times... but F5/compile fixes it... I use this as my "Apply" button for those things now.

**Reply by najak3d** [reply]
Another example of usage -- "LogLevel" - has Debug, Info, Warning, Error --- what if you want to create a new level?   We define LogLevel as an enum -- non-extendible.   If instead it were an OpenEnum, then this list could be extended, and meta data associated with each -- such as special behavior.  I think it's nice to still have it appear as an Enum option, in the code, which OpenEnum allows.

In my mind, the list goes on...  and provides a viable alternative to any place you would:
1. Use an Enum value, but want more meta data, or extendibility
2. OR use a String-Name/Key, but want meta data as well as Enum-style syntax to ensure no typos.

Thank you for helping mature this; sorry it took so long, due to my own noobness with certain things.

**Reply by CyrusNajmabadi** [reply]
> Last nit, is hiding "ReferenceEquals()" - I assume this would require an actual replacement of the existing CompletionProvider.

You could try adding your own hiding signature, and marking it with [EditorBrowsable(never)].  though, if that worked, would only work if you were referenced as a dll, not a source lib.  We don't ever hide legal code from the owners of htat code.

**Reply by CyrusNajmabadi** [reply]
>  I'm surprised that something like this hasn't become a standard part of .NET already (as it is with Java Enums).

File a proposal that explains what it is you're trying to add.  :)

> With OpenEnums - many of these texts can pack more meta data with them, and not just be a string value. For example, text messages might also come with a Style modifier and color code, and "priority level", etc.. So now you can have a whole list of these strings called out like an "enum value" but packed with meta data -- it's no longer "just a string", and yet is as easy to work with as an Enum Value.

I feel like you're justr describing standard OO types :)

**Reply by najak3d** [reply]
@CyrusNajmabadi : "I feel like you're justr describing standard OO types :)"

Yes, exactly-- but usage/access is as though it's an "Enum" value.  It's the marriage between the two, plus adding the ability to extend the enumeration.  There's a reason Java standardized this JavaEnums concept -- it's useful.   

In C# this allows you to have Enums that aren't just:

1.  Integer values only (no added meta data, other than the "name")
2. Non-Extendible.

This is a big deal, IMO.   Most won't know it, until they see it in action, and see their code refactored away from raw String-keys, or raw Enum usage.    All of the devs I've ever met, when they have an Enum Value that needs to be extended -- they simply switch over to String-Keys, and call it done, since they are unaware of a better way to do it in C#.

And until you contributed your time to help me along -- I wasn't aware that this concept could be so elegant so easily within C#.  I had been using a semi-crippled approach, that was still better than raw Strings/Enums, but still lacking.   Thank you for helping to improve this.  I'll be sure to credit you as I push out the improvement.

**Reply by CyrusNajmabadi** [reply]
> Yes, exactly-- but usage/access is as though it's an "Enum" value. 

I genuinely dont' see how that is different.  Having an inheritance hierarchy, and having special statics is just normal C# coding :)   It would help if your proposal actually demonstrated waht this provides on top of normal types taht warrants needing anything special here.

> In C# this allows you to have Enums that aren't just:

Right.  But you can get this today without anything special.

>  There's a reason Java standardized this JavaEnums concept -- it's useful.

No.  You have this a bit backwards.  JavaEnums exist to specify a *fixed* set of values.  They are *non-extendible* intentionally so you have complete control over the entire set of values that are allowed.

You can't say "there's a reason for JavaEnums", and then get rid of the very aspect of their design that justifies their existence.  If you want values in an unbounded fashion, then that is trivial (in both languages).

**Reply by CyrusNajmabadi** [reply]
>This is a big deal, IMO. Most won't know it, until they see it in action, and see their code refactored away from raw String-keys, or raw Enum usage.

I agree.  Using strong types over loose types can often be beneficial.  But, as you've seen, you literally don't even need special code.  You're just utilizing normal C# patterns that have been possible and documented and understood since literally teh beginning :)

**Reply by najak3d** [reply]
> > This is a big deal, IMO. Most won't know it, until they see it in action, and see their code refactored away from raw String-keys, or raw Enum usage.
> 
> I agree. Using strong types over loose types can often be beneficial. But, as you've seen, you literally don't even need special code. You're just utilizing normal C# patterns that have been possible and documented and understood since literally teh beginning :)

That's part of the good news, that implementation was easy.  But in truth, I don't know ANYONE else who actually uses it, nor even knew about the option.    Statically defined string values (or constants) is the ONLY thing I've seen in the past, along with rampant use of hard-coded string values typed in-line.

I'm glad that we've worked through some kinks here, to make the "OpenEnum" base class into something that I think should "sell easily", without resistance.   My StructEnums came with more complexity/caveats, and I was unaware of the "completionlist" tag which alleviates one of the main annoyances that I've lived with for years.

**Reply by CyrusNajmabadi** [reply]
> That's part of the good news, that implementation was easy. But in truth, I don't know ANYONE else who actually uses it, nor even knew about the option. 

I genuinely don't know what to tell you wrt this. :)

**Reply by najak3d** [reply]
How about this -- you've been on the C# lang site and here for many years -- yet this concept seemed "new to you".   So not sure if you are implying "many others are already doing it, and you just don't know about it" ... or that "others aren't using it because they don't see the benefit". ??      My point is that I think they aren't using it, because they just didn't "think of it" and so just defaulted to using the easiest concept -- which is just raw strings: "Heck, raw string keys work; they're extendible/flexible; so why not just use them?  AND... there isn't a better option available in C#."  (many would argue)

**Reply by CyrusNajmabadi** [reply]
> yet this concept seemed "new to you".

It's still new in that you haven't really explained what it is. And as we peel back layers it's not getting clearer. :-)

As you explain more that's when the head scratching has happened and basic questions like "aren't you just describing a class?" come up. 

>  My point is that I think they aren't using it, because they just didn't "think of it" and so just defaulted to using the easiest concep if they're not thinking of inheritance and values, it's unclear why adding another concept entirely would change things for them.
